webrtc/README.md

```diff
@@ -0,0 +1,317 @@
+
+4. _A_ establishes the connection to _B_. _A_ generates a random string and uses
+   that string as the username (_ufrag_ or _username fragment_) and password in
+   the initial STUN message from _A_ to _B_. The random string can be used by
+   _B_ to identify the connection, i.e. demultiplex incoming UDP datagrams per
+   incoming connection. _B_ uses the same random string for the username and
+   password of the STUN message from _B_ to _A_.
+
+5. _B_ does not know the TLS fingerprint of _A_. _B_ upgrades the incoming
+   connection from _A_ as an _insecure_ connection, learning _A_'s TLS
+   fingerprint through the WebRTC DTLS handshake. At this point the DTLS
+   handshake provides confidentiality and integrity but not authenticity.
+
+6. See [Connection Security](#connection-security).
+
+7. See [Multiplexing](#multiplexing).
+
+#### Open Questions
+
+- Is the fact that the server accepts STUN messages from the client prone to
+  attacks?
+
+    - Can an attacker launch an **amplification attack** with the STUN endpoint
+      of the server?
+
+      The QUIC protocol defends against amplification attacks by requiring:
+
+      > an endpoint MUST limit the amount of data it sends to the unvalidated
+      > address to three times the amount of data received from that address.
+
+      https://datatracker.ietf.org/doc/html/rfc9000#section-8
+
+      For WebRTC in libp2p one could require the client (_A_) to add additional
+      bytes to its STUN message, e.g. in the STUN username and password, thus
+      making an amplification attack less attractive.
+
+    - Can a client run a **DOS attack** by sending many STUN messages with
+      different ufrags using different UDP source ports, forcing the server to
+      allocate a new peer connection for each? Would rate limiting suffice to
+      defend against this attack?
+
+  See also:
+  - https://datatracker.ietf.org/doc/html/rfc5389#section-16.2.1
+  - https://datatracker.ietf.org/doc/html/rfc5389#section-16.1.2
+
+- Do the major WebRTC server implementations support using the same UDP port for
+  multiple WebRTC connections, thus requiring multiplexing multiple WebRTC
+  connections on the same UDP port? In particular this came up for NodeJS.
+
+  This is related to [ICE Lite](https://www.rfc-editor.org/rfc/rfc5245), having
+  a host only advertise a single address, namely the host address, which is
+  assumed to be public.
+
+- Do the major (Go / Rust / ...) WebRTC implementations allow us to accept a
+  WebRTC connection from a remote node without previously receiving an SDP
+  packet from such host?
+
+- Can _Browser_ generate a _valid_ SDP packet for the remote node based on the
+  remote's Multiaddr, where that Multiaddr contains the IP, UDP port and TLS
+  certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`)? _Valid_ in
+  the sense that this generated SDP packet can then be used to establish a
+  WebRTC connection to the remote.
+
```

Comment:
@mxinden Currently, the RTCPeerConnection API does not allow setting ICE credentials (ufrag and pwd) directly on the peerconnection. We modify the SDP to change these parameters. While this is disallowed by the spec, we have tested that this works on Safari, Firefox, and Chrome. SDP munging seems to be used in a large number of cases so the current implementations only disallow modification of the fingerprint (refer: https://bugs.chromium.org/p/chromium/issues/detail?id=823036). This is consistent with our earlier testing on the browser.

Reply:
Thanks for raising this @ckousik. I documented this in 774cd52. Let me know in case I am missing something.

Reply:
With the above commit, I will mark this conversation as resolved. Thanks for the input @ckousik.